src/sonic-py-common/sonic_py_common/security_cipher.py

```diff
@@ -0,0 +1,120 @@
+'''
+
+A common module for handling the encryption and
+decryption of the feature passkey. It also takes
+care of storying the secure cipher at root 
+protected file system
+
+'''
+
+import subprocess
+import threading
+import syslog
+import os
+from swsscommon.swsscommon import ConfigDBConnector
+
+class security_cipher:
+    _instance = None
+    _lock = threading.Lock()
+
+    def __new__(cls):
+        with cls._lock:
+            if cls._instance is None:
+                cls._instance = super(security_cipher, cls).__new__(cls)
+                cls._instance._initialized = False
+        return cls._instance
+
+    def __init__(self):
+        if not self._initialized:
+            self._file_path = "/etc/cipher_pass"
+            self._config_db = ConfigDBConnector()
+            self._config_db.connect()
+            # Note: Kept 1st index NA intentionally to map it with the cipher_pass file
+            # contents. The file has a comment at the 1st row / line
+            self._feature_list = ["NA", "TACPLUS", "RADIUS", "LDAP"]
```

Comment:
Do we need a master key per feature? I thought we're using one master key for all features.

Reply:
I guess it is better to have different Master keys for different feature. This way, there will not be any inter-dependency between them.

Reply:
Let's not complicate things for the user, just follow well established practices (e.g a master key for all features) from the popular NOS if possible, my 2 cents.

Reply:
With this approach, we can further extend this implementation for other modules too not just TACPLUS, RADIUS and LDAP. Additionally, it is upto the user if he / she needs to use different keys or can use same keys as well.

Reply:
It will give flexibility to the user, in my opinion.

Reply:
If we go with single key, it will be having huge dependancy in case of changing that key. User needs to change the encrypted passkey in CONFIG_DB for all the features.

Reply:
I disagree, IMO, we should just look at the behavior of other popular NOS that's been there for many years for now, if required for any SONiC use-case, we can think about providing the flexibility in the future.

Reply:
With this approach, we are already providing the flexibility to the user (why to wait for the future :-)). They can have either same or different keys for different features.

The proprietary NOSes have different architectures and they have implemented the feature which can be fitted into their infrastructure. :-) I guess, it is always better if we can add bit of flexibility into any of the designs.